<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Promise最基础的版本--仅支持同步</title>
</head>

<body>
  <script>
    // 判断变量否为function
    const isFunction = variable => typeof variable === 'function'
    // 定义Promise的三种状态常量
    const PENDING = 'PENDING'
    const FULFILLED = 'FULFILLED'
    const REJECTED = 'REJECTED'
    class MyPromise {
      constructor(handle) {
        if (!isFunction(handle)) {
          throw new Error('MyPromise must accept a function as a parameter')
        }
        // 添加状态
        this._status = PENDING
        // 添加状态
        this._value = undefined
        // 执行handle (resolve,reject)=>{},参数：_resolve,_reject
        try {
          handle(this._resolve.bind(this), this._reject.bind(this))
        } catch (err) {
          this._reject(err)
        }
      }
      // 添加resovle时执行的函数 ----------改变status的状态和value的值
      _resolve(val) {
        if (this._status !== PENDING) return
        this._status = FULFILLED
        this._value = val
      }
      // 添加reject时执行的函数
      _reject(err) {
        if (this._status !== PENDING) return
        this._status = REJECTED
        this._value = err
      }
      // 添加then方法
      then(onFulfilled, onRejected) {
        const { _value, _status } = this
        // 返回一个新的Promise对象
        return new MyPromise((onFulfilledNext, onRejectedNext) => {
          switch (_status) {
            case FULFILLED:
              let res = onFulfilled(_value);
              onFulfilledNext(res)
              break
            case REJECTED:
              let err = onRejected(_value);
              onRejectedNext(err)
              break
          }
        })
      }
      // 添加catch方法
      catch(onRejected) {
        return this.then(undefined, onRejected)
      }
    }
  </script>
  <script>
    let p1 = new MyPromise((resolve, reject) => {
      resolve('成功1')
    })
    let p2 = p1.then(res => { throw new Error('error') })
    p2.then(res => { console.log(res, 'success2'); }, err => { console.log(err, 'fail'); })
  </script>
</body>

</html>
